### DISCUSSION TITLE
Recommended way to handle `TcpListener::accept()` errors

### DISCUSSION THREAD

**Discussion by sammko** [hyperium/hyper#3597] [Q&A]
Examples such as `hello.rs` simply propagate any errors returned by `accept()`

https://github.com/hyperium/hyper/blob/9c438e386cbffe17bbbb2b001152ded640e7332e/examples/hello.rs#L44

The Linux `accept(2)` man page states the following:
>    **Error handling**
>        Linux accept() (and accept4()) passes already-pending network
>        errors on the new socket as an error code from accept().  This
>        behavior differs from other BSD socket implementations.  For
>        reliable operation the application should detect the network
>        errors defined for the protocol after accept() and treat them
>        like `EAGAIN` by retrying.  In the case of TCP/IP, these are
>        `ENETDOWN`, `EPROTO`, `ENOPROTOOPT`, `EHOSTDOWN`, `ENONET`, `EHOSTUNREACH`,
>        `EOPNOTSUPP`, and `ENETUNREACH`.

This suggests that at least some return codes of accept should be ignored and accept should be called again. The Rust documentation on `TcpListener::accept` (either std or tokio) is quite scant. I have found [discussions](https://users.rust-lang.org/t/how-to-handle-errors-when-accepting-connections/70838) with no conclusions.

Hyper v0.14 Server had some logic to differentiate accept Errors, but I believe this served a different purpose.
https://github.com/hyperium/hyper/blob/dedcb674f35eaec765a42b550caabe6f694d86d1/src/server/tcp.rs#L279-L286

Is something like this correct or can it get stuck in a busy loop if something is wrong with the socket?
```rust
let stream = match listener.accept().await {
    Ok((stream, _)) => stream,
    #[cfg(target_os = "linux")]
    Err(e) => match e.raw_os_error() {
        Some(
            // Refer to accept(2)
            libc::EAGAIN
            | libc::ENETDOWN
            | libc::EPROTO
            | libc::ENOPROTOOPT
            | libc::EHOSTDOWN
            | libc::ENONET
            | libc::EHOSTUNREACH
            | libc::EOPNOTSUPP
            | libc::ENETUNREACH,
        ) => continue, // Should we sleep here?
        _ => return Err(anyhow!(e).context("Fatal accept error")),
    },
    #[cfg(not(target_os = "linux"))]
    Err(e) => return Err(anyhow!(e).context("Accept error")),
};
```

Does anyone have examples of Rust software where this is handled?

**Comment by sammko** [comment]
Or is this entirely handled by libc? `accept(3)` doesn't really talk about this.

**Comment by beroal** [comment]
See also https://utcc.utoronto.ca/~cks/space/blog/unix/AcceptErrnoProblem and https://stackoverflow.com/questions/76955978/which-socket-accept-errors-are-fatal .
